From a5417ffd03c0fa2041586f3297572d57ee6dd7f5 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Mon, 25 Jan 2021 18:33:09 +0800
Subject: [PATCH 44/81] modesetting: Support multi screens

The output's possible_crtcs should match the screen's available crtcs.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 .../drivers/modesetting/drmmode_display.c     | 30 +++++++++++++++++--
 .../drivers/modesetting/drmmode_display.h     |  1 +
 2 files changed, 28 insertions(+), 3 deletions(-)

diff --git a/hw/xfree86/drivers/modesetting/drmmode_display.c b/hw/xfree86/drivers/modesetting/drmmode_display.c
index 06337f196..5bc4d76b7 100644
--- a/hw/xfree86/drivers/modesetting/drmmode_display.c
+++ b/hw/xfree86/drivers/modesetting/drmmode_display.c
@@ -2449,6 +2449,7 @@ drmmode_crtc_init(ScrnInfoPtr pScrn, drmmode_ptr drmmode, drmModeResPtr mode_res
 {
     xf86CrtcPtr crtc;
     drmmode_crtc_private_ptr drmmode_crtc;
+    xf86CrtcConfigPtr config = XF86_CRTC_CONFIG_PTR(pScrn);
     modesettingEntPtr ms_ent = ms_ent_priv(pScrn);
     drmModeObjectPropertiesPtr props;
     static const drmmode_prop_info_rec crtc_props[] = {
@@ -2458,6 +2459,21 @@ drmmode_crtc_init(ScrnInfoPtr pScrn, drmmode_ptr drmmode, drmModeResPtr mode_res
         [DRMMODE_CRTC_GAMMA_LUT_SIZE] = { .name = "GAMMA_LUT_SIZE" },
         [DRMMODE_CRTC_CTM] = { .name = "CTM" },
     };
+    int o, found = 0;
+
+    /* only init possible crtcs for the outputs */
+    for (o = 0; o < config->num_output; o++) {
+        xf86OutputPtr output = config->output[o];
+        drmmode_output_private_ptr drmmode_output = output->driver_private;
+
+        if (drmmode_output->possible_crtcs & (1 << num)) {
+            output->possible_crtcs |= 1 << config->num_crtc;
+            found = 1;
+        }
+    }
+
+    if (!found)
+        return 0;
 
     crtc = xf86CrtcCreate(pScrn, &drmmode_crtc_funcs);
     if (crtc == NULL)
@@ -3379,10 +3395,18 @@ drmmode_output_init(ScrnInfoPtr pScrn, drmmode_ptr drmmode, drmModeResPtr mode_r
     output->driver_private = drmmode_output;
     output->non_desktop = nonDesktop;
 
-    output->possible_crtcs = 0;
+    drmmode_output->possible_crtcs = 0;
     for (i = 0; i < koutput->count_encoders; i++) {
-        output->possible_crtcs |= (kencoders[i]->possible_crtcs >> crtcshift) & 0x7f;
+        drmmode_output->possible_crtcs |= kencoders[i]->possible_crtcs & 0x7f;
+    }
+
+    if (!dynamic) {
+        /* would be updated in crtc init */
+        output->possible_crtcs = 0;
+    } else {
+        output->possible_crtcs = drmmode_output->possible_crtcs;
     }
+
     /* work out the possible clones later */
     output->possible_clones = 0;
 
@@ -3407,7 +3431,7 @@ drmmode_output_init(ScrnInfoPtr pScrn, drmmode_ptr drmmode, drmModeResPtr mode_r
                 xf86DrvMsgVerb(pScrn->scrnIndex, X_INFO, 0,
                                "Bind output %d to current crtc %d.\n",
                                drmmode_output->output_id, current_crtc);
-                output->possible_crtcs = (1 << i) >> crtcshift;
+                drmmode_output->possible_crtcs = 1 << i;
                 break;
             }
         }
diff --git a/hw/xfree86/drivers/modesetting/drmmode_display.h b/hw/xfree86/drivers/modesetting/drmmode_display.h
index ff1c9b3d7..2b77658d6 100644
--- a/hw/xfree86/drivers/modesetting/drmmode_display.h
+++ b/hw/xfree86/drivers/modesetting/drmmode_display.h
@@ -277,6 +277,7 @@ typedef struct {
     drmmode_prop_info_rec props_connector[DRMMODE_CONNECTOR__COUNT];
     int enc_mask;
     int enc_clone_mask;
+    int possible_crtcs;
     xf86CrtcPtr current_crtc;
     Atom ctm_atom;
     struct drm_color_ctm ctm;
-- 
2.38.5

